### ISSUE TITLE
create a mesh cache strategy for Tpetra and Kokkos

### ISSUE THREAD

**Issue by ecoon** [amanzi/amanzi#301]
Mesh cache will be an approach for migrating Mesh-based data, or more accurately, everything from a mesh that will be needed by a Kokkos functor (i.e. cell volumes, centroids, also connectivity, etc, see below), onto Kokkos-based data structures.

**Comment by ecoon** [comment]
```
Tpetra::Vector<>
- cell volume  (double)
- face area (double)
- edge length (double)

either: Tpetra::DistObject<AmanziGeometry::Point>  OR Tpetra::MultiVector<double>
- cell centroid (double[mesh->spatial_dimension()])  (currently AmanziGeometry::Point)
- face centroid ( "  )
- face normal ( " )
- edge vector ( " )
- node coordinate

Unknown Data structure  (currently std::vector<std::vector< SOMETHING> >)  (maybe can be CSR?)
- cell_face_ids:  for each cell, LID of faces in the closure
- cell_face_dirs: (same as above, +/- 1)
- cell_edge_ids
- cell_2D_edge_dirs  (check with Konstantin?)

- face_cell_ids:  for each face, LID of the cells on either side of the face (length 1 for boundary faces, 2 for interiors)  (note for boundaries this can be {cell_interior, -1}, as these are LIDs, and our LIDs are 0 ... n_cells.
- face_cell_ptype :  ptype is either Mesh::Parallel_type::OWNED or GHOST.
- face_edge_ids
-face_edge_dirs

**Comment by ecoon** [comment]
Implementation suggestion:
 
- nonmember function in src/state 
- accepts a State and a Mesh
- sets requirements for these things, allocates memory, puts them into State

This means that they can be used by Evaluators in the current interface with no new object.

**Comment by ecoon** [comment]
closed by #314 cleaned up in #349
